File: crawlers/github-trending.js
// 爬虫：GitHub Trending — 纯 API，零 LLM
//
// 用法: node github-trending.js [outputDir]
// 输出: {outputDir}/GitHub-Trending/YYYY-MM-DD.md

const https = require("https");
const fs = require("fs");
const path = require("path");

const CONTENT_DIR = process.argv[2] || path.join(__dirname, "..", "content");
const GITHUB_TOKEN = process.env.GITHUB_TOKEN || "";

function githubApi(urlPath) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: "api.github.com",
      path: urlPath,
      headers: {
        "User-Agent": "wiki-crawler/1.0",
        Accept: "application/vnd.github+json",
      },
    };
    if (GITHUB_TOKEN) {
      options.headers.Authorization = `Bearer ${GITHUB_TOKEN}`;
    }

    https
      .get(options, (res) => {
        let data = "";
        res.on("data", (chunk) => (data += chunk));
        res.on("end", () => {
          if (res.statusCode !== 200) {
            reject(new Error(`GitHub API ${res.statusCode}: ${data}`));
            return;
          }
          resolve(JSON.parse(data));
        });
      })
      .on("error", reject);
  });
}

function formatNumber(n) {
  if (n >= 1000) return (n / 1000).toFixed(1) + "k";
  return String(n);
}

async function main() {
  const today = new Date();
  const dateStr = today.toISOString().slice(0, 10);

  // 取昨天的日期用于筛选
  const yesterday = new Date(today);
  yesterday.setDate(yesterday.getDate() - 1);
  const yesterdayStr = yesterday.toISOString().slice(0, 10);

  console.log(`[GitHub Trending] Fetching repos created after ${yesterdayStr}...`);

  let repos;
  try {
    const data = await githubApi(
      `/search/repositories?q=created:>${yesterdayStr}&sort=stars&order=desc&per_page=15`
    );
    repos = data.items || [];
  } catch (err) {
    console.error("[GitHub Trending] API error:", err.message);
    process.exit(1);
  }

  // 过滤低质量项目
  repos = repos.filter((r) => r.stargazers_count >= 10 && !r.archived);

  console.log(`[GitHub Trending] Found ${repos.length} repos`);

  // 生成 Markdown
  const lines = [
    "---",
    `tags: [GitHub, Trending, ${dateStr}]`,
    "---",
    "",
    `# GitHub Trending — ${dateStr}`,
    "",
    "> 每日热门开源项目精选",
    "",
    "## 今日热门项目",
    "",
  ];

  for (const repo of repos) {
    const lang = repo.language ? ` | ${repo.language}` : "";
    const desc = repo.description
      ? repo.description.replace(/\n/g, " ").slice(0, 200)
      : "";
    lines.push(
      `### [${repo.full_name}](${repo.html_url})`,
      "",
      `- ⭐ ${formatNumber(repo.stargazers_count)} stars today${lang}`,
      `- ${desc}`,
      `- [github.com/${repo.full_name}](${repo.html_url})`,
      ""
    );
  }

  if (repos.length === 0) {
    lines.push("> 今日暂无热门项目数据", "");
  }

  lines.push(
    "---",
    "",
    `> 数据来源: [GitHub Search API](https://api.github.com/search/repositories) · 更新时间: ${dateStr}`
  );

  // 写入文件
  const outputDir = path.join(CONTENT_DIR, "GitHub-Trending");
  fs.mkdirSync(outputDir, { recursive: true });
  const outputPath = path.join(outputDir, `${dateStr}.md`);
  fs.writeFileSync(outputPath, lines.join("\n"), "utf8");
  console.log(`[GitHub Trending] Written: ${outputPath}`);
}

main().catch((err) => {
  console.error("[GitHub Trending] Fatal:", err);
  process.exit(1);
});

File: scripts/github_trending.py
#!/usr/bin/env python3
"""
Fetch GitHub Trending-like repository data with the GitHub Search API.

This script intentionally uses only the Python standard library so it can run
inside GitHub Actions without extra setup.
"""

import argparse
import json
import os
import sys
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone


PERIOD_DAYS = {"daily": 1, "weekly": 7, "monthly": 30}
LANGUAGES = ["", "python", "javascript", "typescript", "go", "rust", "java"]

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8")


def github_search(query: str, per_page: int, token: str) -> list[dict]:
    params = urllib.parse.urlencode(
        {
            "q": query,
            "sort": "stars",
            "order": "desc",
            "per_page": min(per_page, 100),
        }
    )
    req = urllib.request.Request(
        f"https://api.github.com/search/repositories?{params}",
        headers={
            "Accept": "application/vnd.github+json",
            "User-Agent": "otae-wiki-trending/1.0",
            "X-GitHub-Api-Version": "2022-11-28",
            **({"Authorization": f"Bearer {token}"} if token else {}),
        },
    )

    try:
        with urllib.request.urlopen(req, timeout=25) as resp:
            return json.loads(resp.read()).get("items", [])
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        print(f"[WARN] GitHub API HTTP {exc.code}: {body[:240]}", file=sys.stderr)
    except Exception as exc:
        print(f"[WARN] GitHub API error: {exc}", file=sys.stderr)
    return []


def fetch_trending(period: str, limit: int, language: str, token: str) -> list[dict]:
    since = (datetime.now(timezone.utc) - timedelta(days=PERIOD_DAYS[period])).strftime(
        "%Y-%m-%d"
    )
    seen: set[str] = set()
    repos: list[dict] = []

    languages = [language] if language else LANGUAGES
    for lang in languages:
        if len(repos) >= limit * 2:
            break
        lang_filter = f" language:{lang}" if lang else ""
        min_stars = 10 if lang else 50
        query = f"pushed:>={since} stars:>={min_stars}{lang_filter}"
        for item in github_search(query, per_page=limit * 2, token=token):
            name = item.get("full_name")
            if name and name not in seen:
                seen.add(name)
                repos.append(item)

    repos.sort(key=lambda repo: repo.get("stargazers_count", 0), reverse=True)
    return repos[:limit]


def main() -> None:
    parser = argparse.ArgumentParser(description="Fetch GitHub trending repositories")
    parser.add_argument("--period", choices=PERIOD_DAYS.keys(), default="daily")
    parser.add_argument("--limit", type=int, default=25)
    parser.add_argument("--language", default="")
    parser.add_argument("--token", default=os.environ.get("GITHUB_TOKEN", ""))
    parser.add_argument("--output", default="", help="Write JSON output to this file")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    print(f"[INFO] Fetching GitHub trending: {args.period}", file=sys.stderr)
    repos = fetch_trending(args.period, args.limit, args.language, args.token)
    if not repos:
        print("[ERROR] No repositories fetched", file=sys.stderr)
        sys.exit(1)

    output = [
        {
            "rank": index,
            "name": repo["full_name"],
            "url": repo["html_url"],
            "description": repo.get("description") or "",
            "stars": repo.get("stargazers_count", 0),
            "forks": repo.get("forks_count", 0),
            "language": repo.get("language") or "",
            "topics": repo.get("topics", []),
        }
        for index, repo in enumerate(repos, 1)
    ]

    if args.json:
        if args.output:
            with open(args.output, "w", encoding="utf-8") as fp:
                json.dump(output, fp, ensure_ascii=False, indent=2)
        else:
            json.dump(output, sys.stdout, ensure_ascii=False, indent=2)
    else:
        for repo in output:
            print(f"#{repo['rank']} {repo['name']} {repo['stars']} stars")


if __name__ == "__main__":
    main()
